### ISSUE TITLE
CLI should not catch all exceptions

### ISSUE THREAD

**Issue by piotrpMSFT** [dotnet/cli#5076]
CLI uses exceptions to send user-friendly messages when user-correctable error conditions are identified. The exceptions are caught close to the CLI's `Main` and are printed to the console.

The catch clause needs to be constrained to only handle exceptions intended for this scenario. It should not catch all exceptions. When the CLI throws an exception due to a bug in the CLI itself, the exception should at least print a full stack trace to improve diagnosability, ideally with instructions to report the issue at the CLI's repo.

@nguerrera does this sound reasonable to you? Or do you think we should let the exception go and let OS error reporting mechanisms facilitate the feedback loop?

**Comment by nguerrera** [comment]
> @nguerrera does this sound reasonable to you?

Yes.

>  Or do you think we should let the exception go and let OS error reporting mechanisms facilitate the feedback loop?

+@jaredpar about reporting mechanism. It would be nice to have parity with the compiler with respect to getting automated reports with a crash dump, etc.

**Comment by jaredpar** [comment]
The command line compiler just lets unexpected exceptions go unhandled and relies on the OS reporting mechanisms.

**Comment by krwq** [comment]
@piotrpMSFT @livarcocc despite the stuff in dotnet.dll I'm seeing some migration code also choking many exceptions which likely should not be (haven't investigated all those cases). Should this be done as part of that or is it ok for migration code to do that? (I consider it legacy code so IMO we should not change it)

Here is the complete list of stuff which catches all exceptions:
```
dotnet\commands\dotnet-migrate\Program.cs:81:            catch (Exception ex)
dotnet\commands\dotnet-msbuild\MSBuildForwardingApp.cs:49:                catch (Exception)
dotnet\commands\dotnet-msbuild\MSBuildLogger.cs:29:            catch (Exception)
dotnet\commands\dotnet-msbuild\MSBuildLogger.cs:49:            catch(Exception)
dotnet\commands\dotnet-msbuild\MSBuildLogger.cs:66:            catch (Exception)
dotnet\commands\dotnet-new\Program.cs:213:            catch (Exception ex)
dotnet\MulticoreJitActivator.cs:54:            catch (Exception)
dotnet\Program.cs:81:            catch (Exception ex)
dotnet\Telemetry.cs:97:            catch (Exception)
dotnet\Telemetry.cs:120:            catch (Exception)
dotnet-archive\Program.cs:97:            catch (Exception ex)
Microsoft.DotNet.Cli.Utils\CommandResolution\ProjectToolsCommandResolver.cs:316:            catch (Exception e)
Microsoft.DotNet.Cli.Utils\CommandResolution\ProjectToolsCommandResolver.cs:326:                catch (Exception e2)
Microsoft.DotNet.Configurer\NuGetCacheSentinel.cs:95:            catch { }
Microsoft.DotNet.InternalAbstractions\FileWrapper.cs:47:            catch { }
Microsoft.DotNet.InternalAbstractions\TemporaryDirectory.cs:25:            catch

### Migration specific
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\Files\PatternsCollectionHelper.cs:46:            catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\GlobalSettings.cs:54:            catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\ProjectReader.cs:57:            catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\ProjectReader.cs:121:                catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\ProjectReader.cs:352:                        catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\ProjectReader.cs:476:                    catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\Utilities\ResilientFileStreamOpener.cs:31:                catch
Microsoft.DotNet.ProjectJsonMigration\Microsoft.DotNet.Internal.ProjectModel\Workspace.cs:175:                        catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\ProjectDependencyFinder.cs:433:                catch (Exception ex)
Microsoft.DotNet.ProjectJsonMigration\ProjectMigrator.cs:90:            } catch {}
Microsoft.DotNet.ProjectJsonMigration\ProjectMigrator.cs:97:                } catch { }
```

**Comment by piotrpMSFT** [comment]
Fixed by #5206
